src/mechanics/camera.rs

```diff
@@ -1,18 +1,489 @@
 ) {
-    if let Ok(player_transform) = player_query.get_single() {
-        if let Ok(mut camera_transform) = camera_query.get_single_mut() {
-            let new_camera_position = Vec3::new(
-                player_transform.translation.x,
-                player_transform.translation.y,
-                camera_transform.translation.z,
-            );
-            camera_transform.translation = new_camera_position;
-        }
+    if camera_query.is_empty() {
+        return;
+    }
+
+    if player_query.is_empty() {
+        return;
+    }
+
+    if level_dimension.height == 0 || level_dimension.width == 0 {
+        return;
+    }
+
+    let (mut camera_transform, camera_bounds) = camera_query.get_single_mut().unwrap();
+    let player_transform = player_query.get_single().unwrap();
+    
+    let camera_width = (camera_bounds.right - camera_bounds.left).abs() + 1.0;
+    let camera_height = (camera_bounds.top - camera_bounds.bottom).abs() + 1.0;
+
+    if camera_width > level_dimension.width as f32 {
+        camera_transform.translation.x = level_dimension.width as f32 / 2.0;
+    }
+    else {
+        camera_transform.translation.x = player_transform.translation.x.clamp(camera_width / 2.0, level_dimension.width as f32 - (camera_width / 2.0));
+    }
+
+    if camera_height > level_dimension.height as f32 {
+        camera_transform.translation.y = level_dimension.height as f32 / 2.0;
+    }
+    else {
+        camera_transform.translation.y = player_transform.translation.y.clamp(camera_height / 2.0, level_dimension.height as f32 - (camera_height / 2.0));
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const TEST_LEVEL_WIDTH: usize = 500;
+    const TEST_LEVEL_HEIGHT: usize = 500;
+
+    const CAMERA_HEIGHT: usize = 100;
+    const CAMERA_WIDTH: usize = 100;
+
+    const CAMERA_MIDPOINT: usize = 50;
+
+    const TEST_LEVEL_WIDTH_IN_BOUNDS: f32 = 250.0;
+    const TEST_LEVEL_WIDTH_OUT_LBOUNDS: f32 = -500.0;
+    const TEST_LEVEL_WIDTH_OUT_RBOUNDS: f32 = 1000.0;
+
+    const TEST_LEVEL_HEIGHT_IN_BOUNDS: f32 = 250.0;
+    const TEST_LEVEL_HEIGHT_OUT_TBOUNDS: f32 = 1000.0;
+    const TEST_LEVEL_HEIGHT_OUT_BBOUNDS: f32 = -500.0;
+
+    fn setup_app_bounds_checking() -> App {
+        let mut app = App::new();
+
+        app.insert_resource(LevelDimensions {
+            width: TEST_LEVEL_WIDTH,
+            height: TEST_LEVEL_HEIGHT,
+        });
+
+        app.add_system(move_camera);
+
+        app
+    }
+
+    #[test]
+    fn within_bounds() {
+        let mut app = setup_app_bounds_checking();
+
+        let _player_id = app
+            .world
+            .spawn()
+            .insert_bundle((
+                Player,
+                Transform::from_xyz(
+                    TEST_LEVEL_WIDTH_IN_BOUNDS, 
+                    TEST_LEVEL_HEIGHT_IN_BOUNDS,
+                    0.0),
+            ))
+            .id();
```

Comment:
If you just want to spawn the player into the world without tracking it for later, what do you think about changing it to just this then?
```rust
// The camera's position is dependent off of the Player's position whenever it changes, so we need to
// spawn the Player to trigger the camera to move.
app.world.spawn().insert_bundle((
    Player,
    Transform::from_xyz(TEST_LEVEL_WIDTH_IN_BOUNDS, TEST_LEVEL_HEIGHT_IN_BOUNDS, 0.0),
));
```

Reply:
Good idea!